src/vanilla/store.ts

```diff
@@ -198,8 +194,14 @@ const registerBatchAtom = (
 ) => {
   if (!batch.D.has(atom)) {
     batch.D.set(atom, new Set())
-    addBatchFunc(batch, 'M', () => {
-      atomState.m?.l.forEach((listener) => addBatchFunc(batch, 'M', listener))
+    addBatchFunc(batch, 'H', () => {
+      for (const listener of atomState.m?.l || []) {
+        let priority: BatchPriority = 'M'
+        if ('INTERNAL_priority' in listener) {
+          priority = listener.INTERNAL_priority as BatchPriority
```

Comment:
I couldn't think of a way to register a high priority listener without adding a new api, so I went with an `INTERNAL_priority` property on the listener function itself. I don't think we should include this in the typescript definition though, because I don't really think there's a use-case outside of sync effects.

Reply:
Update: I refactored this in a later commit. Now INTERNAL_onInit returns the store and the atomState. There is a new property on the atomState for registering additional event listeners.